### ISSUE TITLE
CKAN 2.9 upgrade

### ISSUE THREAD

**Issue by chris48s** [okfn/ckanext-unhcr#399]
We're a way off starting this, but I've had a few thoughts swimming around in my mind, so now is a good time to start collecting them in an issue.

There are several categories of work to think about here:

- Filling in gaps in test coverage before we start
- Core patches
  Review core patches in https://github.com/okfn/docker-ckan-unhcr-aws/tree/master/ckan/patches/ckan
  Do we still need them? Do they need to be updated?
- Documented backwards-incompatible api changes to public API
- Un-documented backwards-incompatible changes to internals
  There are a number of places where we're patching/overriding internals - the internals may have changed and our overrides/patches will need to change too. Reviewing places where we import from core is one thing. Working out what bits of templates we're overriding is probably going to be more painful
- Removing customisations: Sometimes we've done something custom to fix a bug/add a feature that wasn't in core, or added it via a plugin. Now it is in core and we can remove the customisation/plugin
- Toolchain updates e.g:
    - Pylons controllers --> Flask blueprints
    - Custom migrations --> Alembic
    - Nose --> Pytest
    - Paster commands --> Click
    - Fanstatic --> Webassets
- Front-end updates (e.g: 2.7 --> 2.8 changed the bootstrap version. Anything like that we need to account for?)
- Plugins. We use:
    - ckanext-pdfview
    - ckanext-hierarchy
    - ckanext-facetcollapse
    - ckanext-collaborators
    - ckanext-scheming
    - ckanext-harvest
    - ckanext-ddi
    - ckanext-saml2
    - ckanext-cloudstorage (fork)
    They all need to be updated. Question: our plan is to go ckan 2.8/python 2 --> ckan 2.9/python 2 --> ckan 2.9/python 3. If a plugin we rely on drops python 2 compatibility in the first 2.9-compatible releases, does that scupper the whole thing?

**Comment by amercader** [comment]
Useful info: https://github.com/ckan/ckan/wiki/Python-3-migration-guide-for-extensions

**Comment by chris48s** [comment]
* In CKAN 2.9 collaborators is a native feature but the implementation differs a bit from ckanext-collaborators so some changes will need to be made around this, particularly for code that interacts a lot with collaborators like the "access requests" concept.
* There are some changes to Activities in CKAN 2.9 which we will need to accommodate in our customisations for activities
* We added a feature in https://github.com/okfn/ckanext-unhcr/pull/364 to managed sysadmins from the web UI. I then also ported this upstream to core in https://github.com/ckan/ckan/pull/5522 It should be possible to remove all of that code, but the implementation we ended up merging in core was a bit different. Does this cause us a problem?

**Comment by chris48s** [comment]
Evaluate switching from https://github.com/okfn/ckanext-saml2 to https://github.com/keitaroinc/ckanext-saml2auth

**Comment by chris48s** [comment]
couple of other random 2.9 bits:

https://github.com/okfn/ckanext-unhcr/pull/400#issuecomment-670501654

https://github.com/okfn/ckanext-unhcr/blob/4225819138e17236388933068ad7b8400197bbb7/ckanext/unhcr/plugin.py#L49-L51

**Comment by chris48s** [comment]
https://github.com/okfn/ckanext-unhcr/pull/459#issuecomment-746362096

**Comment by chris48s** [comment]
Done a bit of a preliminary review of our plugins. Lets kick off with the easy ones...

* [okfn/ckanext-collaborators](https://github.com/okfn/ckanext-collaborators): Collaborators is native in 2.9. Get rid of it
* [ckan/ckanext-scheming](https://github.com/ckan/ckanext-scheming):  Already updated for CKAN 2.9/Python 3 :tada: 
* [ckan/ckanext-harvest](https://github.com/ckan/ckanext-harvest): Already updated for CKAN 2.9/Python 3 :tada: 

These ones are more complicated:

## [ckan/ckanext-pdfview](https://github.com/ckan/ckanext-pdfview)

* Needs upgrading for CKAN 2.9
* There is an open PR for this, but it is not merged. What is needed to get it over the line? https://github.com/ckan/ckanext-pdfview/pull/36
* There is also a WIP proposal to merge PDF Preview (back) into CKAN core. Could this land in a 2.9.x release, or would it be next feature release (2.10? 3.0?) if accepted? https://github.com/ckan/ckan/pull/5541
* Where is the sensible place to focus our efforts?

## [okfn/ckanext-hierarchy](https://github.com/okfn/ckanext-hierarchy)

* We are using our own fork of https://github.com/ckan/ckanext-hierarchy
* https://github.com/ckan/ckanext-hierarchy has been updated for CKAN 2.9/python 3 compatibility: https://github.com/ckan/ckanext-hierarchy/pull/42
* Can we switch to ckan/ckanext-hierarchy? Should we port the upstream changes to our fork? Should we try to contribute our customisations upstream?

## [okfn/ckanext-facetcollapse](https://github.com/okfn/ckanext-facetcollapse/)

* Needs upgrading for CKAN 2.9 (or at least checking/testing to see if it still works)
* This codebase looks like it is pretty small but there are no tests :( https://github.com/okfn/ckanext-facetcollapse/blob/master/ckanext/facetcollapse/tests/test_plugin.py

## [okfn/ckanext-ddi](https://github.com/okfn/ckanext-ddi)

* We are using a custom fork of https://github.com/liip/ckanext-ddi
* It will need checking/upgrading for CKAN 2.9
* This codebase also looks like it will be a pain to work on because it also has no tests at all. It is also a larger codebase than facetcollapse

## [okfn/ckanext-saml2](https://github.com/okfn/ckanext-saml2)

This one is tricky:
* We've talked about evaluating https://github.com/keitaroinc/ckanext-saml2auth as a solution.
* keitaroinc/ckanext-saml2auth is python 3 only. If our plan is to go 2.8/py2 --> 2.9/py2 --> 2.9/py3 we need an intermediate solution for the 2.9/py2 phase
* Upgrade okfn/ckanext-saml2 to work with 2.9? Backport keitaroinc/ckanext-saml2auth to work with python 2?
* There are other projects using one or other of the multiple forks of ckanext-saml2. This is a wider ecosystem problem

## [okfn/ckanext-cloudstorage](https://github.com/okfn/ckanext-cloudstorage)

* We're using our own custom fork of https://github.com/TkTech/ckanext-cloudstorage
* Needs upgrading for CKAN 2.9 (or at least checking/testing to see if it still works)
* No tests, quite a large codebase
* ckanext-cloudstorage has a certain amount of code that supports use-cases we don't care about (e.g: Azure storage) and also some pain points (e.g: https://github.com/okfn/ckanext-unhcr/issues/437 ). Maybe if we need to do a lot of work to upgrade, this is an opportunity to revisit this plugin.
* We might be able to take some inspiration from https://github.com/TkTech/ckanext-cloudstorage/pull/42

**Comment by chris48s** [comment]
I had a look over https://github.com/keitaroinc/ckanext-s3filestore today and installed the CKAN 2.8 version on my local RIDL install. Here's a quick rundown:

- It does support secure URLs
- It uses the same filesystem layout in S3 as okfn/ckanext-cloudstorage does. This means we can switch out cloudstorage for s3filestore and everything will _just work™_. There's no migration process needed
- There's a 2.8/py2 version and a 2.9/py3 version. I've not tested it on 2.9/py2 but looking at the diff between the 2 versions if we need to do something to get it to run on 2.9/py2 as an interim measure it shouldn't be a big deal
- API file uploads work
- All of our customisations in okfn/ckanext-cloudstorage are around trying to make the JS-based multipart upload work better. They're all irrelevant because..
- keitaroinc/ckanext-s3filestore doesn't implement a JS-based multipart upload. It just submits files in the POST request like vanilla CKAN does. You might consider this an advantage or disadvantage, but I think in this context it is actually a good thing. We've explicitly limited the upload size for RIDL https://github.com/okfn/docker-ckan-unhcr-aws/blob/9ba36b46f4cc81e10a8012a40a3cbca21775a887/ckan/environment/base.env#L12 and although we've tried to make some improvements to the CloudStorage plugin I don't think the client has ever really been happy with the error handling/notifications/etc with the JS-based uploader.
- The `resource_download` route in s3filestore is a flask blueprint not a pylons controller so we'll have to modify the changes from https://github.com/okfn/ckanext-unhcr/commit/2a90fc935657a25c5448cd250a2053feba0a1f1b but that's part of the 2.8 --> 2.9 migration anyway. I don't think it is a big deal.

Happy to review the datopian solution too , but I think moving to keitaroinc/ckanext-s3filestore is probably the easy option for us here.

**Comment by chris48s** [comment]
Having spent slightly more time looking at keitaroinc/ckanext-s3filestore the tests were passing on CKAN 2.9/py2 up to https://github.com/keitaroinc/ckanext-s3filestore/commit/9b74c29c019e8a51555db214b58700ffb3a6da90 so we should be good to use `pip install -e git+https://github.com/keitaroinc/ckanext-s3filestore.git@9b74c29c019e8a51555db214b58700ffb3a6da90#egg=ckanext-s3filestore` while we are on CKAN 2.9/py2. It looks like nothing of much significance has been committed since then.

**Comment by chris48s** [comment]
Note: also remember this when it comes to storage plugins: https://github.com/okfn/ckanext-unhcr/pull/459
I think switching to keitaroinc/ckanext-s3filestore shouldn't really change anything there because uploads via the frontend will just use the same process as uploads via the API (and hence fire the same events), but double-check you don't trip over this again.

**Comment by chris48s** [comment]
Gotcha on using `with_plugins` pytest fixture:

> The `with_plugins` fixture has an unexpected behaviour, that after the
> test is run it unloads all the plugins but does not load the default
> ones (the ones on the ini file). This means it needs to be called again
> on each test class. This should probably be fixed in CKAN core.
> 
> Additionally, plugins are only loaded at the start of each test, so
> steps performed in the class methods (setup_class) won't have the
> plugins loaded. This means for example that if we create a dataset using
> a factory it won't use the custom schema because scheming is not loaded.

**Comment by chris48s** [comment]
One thing to particularly keep on top of as we migrate controllers to blueprints will be updating the `url_for()` calls in link targets, form actions, redirects, etc

**Comment by chris48s** [comment]
There are two plugin changes which are tied together here:

- okfn/ckanext-cloudstorage --> keitaroinc/ckanext-s3filestore
- ckanext-pdfview 0.0.5 --> 0.0.6

There are several moving parts that fit together, but I will try to keep to the key points and not ramble on too much.

- I upgraded to ckanext-pdfview==0.0.6 but it doesn't work well with okfn/ckanext-cloudstorage because ckanext-cloudstorage doesn't set the content type header correctly on PDFs
- Given we're planning to move away from okfn/ckanext-cloudstorage (which has other issues) soon as part of the 2.9 migration, I don't see the point in fixing it there
- Good news :+1: keitaroinc/ckanext-s3filestore does have code to detect and set a sensible content-type on upload https://github.com/keitaroinc/ckanext-s3filestore/blob/6dd6a8c68152346638439f400eea24bbc8d88da7/ckanext/s3filestore/uploader.py#L313-L328
- Bad news :-1: keitaroinc/ckanext-s3filestore serves everything with a `Content-Disposition: attachment` header, which also prevents the PDF preview from working. There is a `preview` flag but it is ignored unless the content-type is `text/html` https://github.com/keitaroinc/ckanext-s3filestore/blob/6dd6a8c68152346638439f400eea24bbc8d88da7/ckanext/s3filestore/views/resource.py#L70-L77

so.. here's the plan:

- We stick on ckanext-pdfview==0.0.5 for the moment. The JS-based approach does cope with the PDF being served with both the content type and content disposition headers which are giving us problems with the HTML `<object>` based approach
- We move from okfn/ckanext-cloudstorage to keitaroinc/ckanext-s3filestore and get the client to approve it. This change is more complicated and involves user-facing changes so lets not put time into sending PRs to other projects until we're confident the client is OK with it as a solution. See PRs:
    - https://github.com/okfn/docker-ckan-unhcr-aws/pull/29
    - https://github.com/okfn/ckanext-unhcr/pull/500
      Lets get those reviewed/deployed to UAT and approved as a first step.
- Once we're settled on that and deployed to UAT and Prod:
    - We send some PRs to other projects:
        - A PR to keitaroinc/ckanext-s3filestore (2.9 branch) to make the preview param either ignore mimetype completely or at least work for `application/pdf` too
        - A PR to keitaroinc/ckanext-s3filestore (2.8 branch) backporting that change
        - A PR to ckanext-pdfview to call the resource download endpoint with `?preview=1`
    - We need to run an ad-hoc migration job on both the UAT and prod buckets to find any existing PDF files in the buckets where the content-type is `binary/octet-stream` and update it to `application/pdf` (any new files will have the right type once we switch to s3filestore) but all the ones we uploaded in the past need fixing
- Once we've done all that, we can finally circle back and upgrade ckanext-pdfview to a CKAN 2.9-compatible version :tada:
- We should also propose adding the `?preview=1` flag to https://github.com/ckan/ckan/pull/5541 so that we don't just end up with the same issue again when this eventually lands in core

**Comment by chris48s** [comment]
PDF preview PRs:
https://github.com/keitaroinc/ckanext-s3filestore/pull/2
https://github.com/ckan/ckanext-pdfview/pull/42

**Comment by chris48s** [comment]
I think everything discussed in this issue is either now already done, in progress or I've covered it in one of the more specific issues under the CKAN 2.9 milestone: https://github.com/okfn/ckanext-unhcr/milestone/4 so I am going to close it.
